### PR TITLE
Completing read prompter

### PR DESCRIPTION
**Pull request by minad**
A few more steps towards #186. Unfortunately no Marginalia annotation and no non-general Embark actions for now.

### CODE REVIEW

File: embark.el [hunk 1]
```diff
@@ -692,7 +693,8 @@ first line of the documentation string; otherwise use the word
                   (if (eq action 'metadata)
                       `(metadata (category . command))
                     (complete-with-action action commands string predicate)))
-                nil t nil 'embark--prompter-history))
+                nil t nil 'embark--prompter-history
+                (and def (symbol-name def))))
```
Comment (oantolin) [inline embark.el]: I think it would be better to use embark--comand-name rather than symbol-name, to allow for lambdas, simple menu items, etc.

Comment (oantolin) [reply embark.el]: Ah, this is obsoleted my a later commit, forget it.

**Comment by oantolin** [comment]
It would be a tiny bit of code duplication between here and Marginalia, but I think maybe we should handle the annotations right here (specially since it's simpler than the Marginalia annotator, because we wouldn't include the keybinding).

Then we'd add an transformer that chops off the keybinding and reports type `command`, and we'd be set.

**Comment by oantolin** [comment]
I can merge this and add the annotation and transformer myself.

**Comment by minad** [comment]
> It would be a tiny bit of code duplication between here and Marginalia, but I think maybe we should handle the annotations right here (specially since it's simpler than the Marginalia annotator, because we wouldn't include the keybinding).

I think it is better to handle it in Marginalia, since the function `marginalia--function-doc` is non-trivial.

> Then we'd add an transformer that chops off the keybinding and reports type command, and we'd be set.

Yes. We should do this to regain the actions. Use 'symbol or better use 'command?

**Comment by minad** [comment]
Instead of chopping of the keybinding I would attach a symbol 'embark-command to the formatted string. Then we can use that in the annotator and the transformer.

**Comment by oantolin** [comment]
> I think it is better to handle it in Marginalia, since the function `marginalia--function-doc` is non-trivial.

I had forgotten about skipping the lines of the docstring reporting advice. I agree it would be better to do this in Marginalia then.
 
> Yes. We should do this to regain the actions. Use 'symbol or better use 'command?

Surely `command` is better since these are commands! The `command` category gives you access to all the symbol actions plus binding to a key locally/globally and `Info-goto-emacs-command-node`.

**Comment by oantolin** [comment]
> Instead of chopping of the keybinding I would attach a symbol 'embark-command to the formatted string. Then we can use that in the annotator and the transformer.

OK, yes, that would be better.

**Comment by minad** [comment]
> Instead of chopping of the keybinding I would attach a symbol 'embark-command to the formatted string. Then we can use that in the annotator and the transformer.

When attaching the 'embark-command we should also handle menu items better, such that we do not confuse the desc with the cmd of `(desc . cmd)`.

**Comment by oantolin** [comment]
Do you want to do it before I merge? Or should I merge and then do the rest myself? Whatever you want.

**Comment by oantolin** [comment]
> When attaching the 'embark-command we should also handle menu items better, such that we do not confuse the desc with the cmd of `(desc . cmd)`. Then the `consult-narrow` bindings also get a Marginalia annotation.

What do you mean exactly. I imagined the `embark-command property being set to `cmd` in the `(desc . cmd)` case.

**Comment by minad** [comment]
> What do you mean exactly? I imagined the embark-command property being set to cmdin the(desc . cmd)` case.

Yes, I mean exactly that.

> Do you want to do it before I merge? Or should I merge and then do the rest myself? Whatever you want.

If you have time now, feel free to take over.

**Comment by oantolin** [comment]
> If you have time now, feel free to take over.

OK, will do! Thanks for the changes you made! One tiny thing I liked is not taking the `cdr` of the result of `assoc`, might as well reduce indentation and handle it in the `pcase`. I keep the lines pretty short in Embark because I often don't maximize my Emacs.

**Comment by minad** [comment]
I feel the embark completing read prompter is really a neat missing piece. We often had these requests before for `consult-descbinds` and I never understood why one would want this (I still don't), but using this as `prefix-help-command` is just the perfect fit. Maybe we should still rename the function to `embark-bindings`, make it a command and pass the current keymap as default to cater for the `consult-descbinds` needs?

**Comment by oantolin** [comment]
I added the transformer. a103cfdf311167835145b236be4c9b40d700af6c And will now add the annotator to Marginalia.

**Comment by oantolin** [comment]
Do you think in Marginalia the annotator should be considered light or heavy?

**Comment by minad** [comment]
> Do you think in Marginalia the annotator should be considered light or heavy?

Hmm, I think it is a heavy one, since for commands we generally don't add docstrings. 

Maybe the light/heavy idea was a bad one, it seems everyone is configuring Marginalia to just use the heavy annotators :laughing:

**Comment by oantolin** [comment]
I wouldn't be against removing the light/heavy distinction. Do people cycle? Maybe that whole thing is overkill. Maybe there should be a single alist of annotators. If people want cycling, they can write a command that swaps out the entire alist.

**Comment by oantolin** [comment]
> Hmm, I think it is a heavy one, since for commands we don't add docstrings in the light variant.

The keybindings buffer doesn't have docstrings either, and this for a replacement of it, so I guess that also points to it being heavy.

**Comment by oantolin** [comment]
`when-let` can handle a single set of parens, `(when-let (a b) ...)` rather than `(when-let ((a b)) ...)`? Weird. I'll use it to blend in.

**Comment by minad** [comment]
> Do people cycle? 

Well, I cycle - when testing the heavy annotators since I am stubbornly still using the light annotators. But after I added `x-group-function` the heavy annotators are less annoying since the ugly misalignment in `consult-buffer` is nicely separated by the grouping lines.

> Maybe there should be a single alist of annotators. If people want cycling, they can write a command that swaps out the entire alist.

Yes, I am considering this.

**Comment by oantolin** [comment]
OK, done. minad/marginalia@f3eaa1a0372c54295a3ae9b800527d58c991ff3d.

**Comment by oantolin** [comment]
You've been handling Marginalia much more than I have, feel free to change names or whatever, if I didn't get the style right.

**Comment by oantolin** [comment]
> I feel the embark completing read prompter is really a neat missing piece. We often had these requests before for `consult-descbinds` and I never understood why one would want this (I still don't), but using this as `prefix-help-command` is just the perfect fit. Maybe we should rename the `embark-prefix-help-command` to `embark-bindings` and support the case where the command is called without prefix to cater for the `consult-descbinds` needs.

Yes, it's great as a value for `prefix-help-command`. I'm not sure I'd ever run it from top level, it seems that that's what `M-x` is for... 😆 

I'd be OK with the renaming, we'd just need to figure out a good way to tell if it has been called from a prefix or not.

**Comment by minad** [comment]
> Yes, it's great as a value for prefix-help-command. I'm not sure I'd ever run it from top level, it seems that that's what M-x is for... laughing

I think the point is to explore keybindings. That's valuable additionally to M-x.

**Comment by oantolin** [comment]
Say your M-x shows keybindings (maybe you're a Marginalia user or you use the `*Completions*` buffer). Then the value in a descbinds command is that it skips commands that are unbound? Oh, and that you can match against the bindings (which seems like a weird thing to want to do). I'm not sure I see the point, but OK.

**Comment by minad** [comment]
I thought a bit about the light/heavy annotators. One point is also that you may want to use Embark *without annotators*. This only works if you set `marginalia-annotators` to `'(nil ...)`. So there is some value in the indirection. 

The light/heavy distinction could be considered kind of unnecessary however. But I must also say that I never missed heavy annotations when running `consult-buffer`, `find-file` or `M-x`, since for these commands I am usually doing a quick action, and not some kind of exploration. I think the situation could change if we get https://github.com/oantolin/orderless/issues/30, such that I could for example match all elisp buffers via `@elisp`. Then I may want to see the buffer metadata always.

**Comment by minad** [comment]
> Then the value in a descbinds command is that it skips commands that are unbound? Oh

Yes.

> Oh, and that you can match against the bindings (which seems like a weird thing to want to do).

Yes.

> I'm not sure I see the point, but OK.

I also don't see it. In the context of Consult I argued that one should rather use `describe-bindings` + `consult-focus-lines` to achieve a similar filtering. But in combination with the Marginalia annotators we get a little more.

For me the main point now is that we get the functionality basically for free since we already have the `embark-prefix-help-command`. At this point I think it is just worth to have this since the cost is so low. Having it as a separate consult command didn't make sense. But here we already have all the pieces!

**Comment by oantolin** [comment]
> At this point I think it is just worth to have this since the cost is so low. Having it as a separate consult command didn't make sense. But here we already have all the pieces!

Agreed.

**Pull request merged**